[FLINK-19743] Add metric definitions in FLIP-33 and report some of them by default.

becketqin opened the pull request:
## What is the purpose of the change
This patch does the following:
* Moved the `MetricNames` from flink-runtime to flink-core and added the metrics defined in FLIP-33.
* Improved the `SourceReaderBase` API to allow additional information to be passed from the `SplitReader` to `RecordEmitter`. So that the implementations do not have to put some common information in the records themselves.
* Report some of the metrics by default.

Metrics implemented by the SourceOperator / ReaderOutput:
1. currentEmitEventTimeLag (at a sampling frequency of once per second)
2. watermarkLag

Metrics implemented in the `SourceReaderBase` (can be overridden by individual implementations):
1. numRecordsIn / numRecordsInPerSecond
2. sourceIdleTime

Metrics left for individual implementations to report in `SplitReader` / `RecordEmitter` :
1. numBytesIn / numBytesInPerSecond
2. currentFetchEventTime
3. pendingBytes
4. pendingRecords
5. numRecordsInErrors (errors in SplitReader and record emitter)

## Brief change log
* Moved the `MetricNames` from flink-runtime to flink-core and added the metrics defined in FLIP-33.
* Improved the `SourceReaderBase` API to allow additional information to be passed from the `SplitReader` to `RecordEmitter`. So that the implementations do not have to put some common information in the records themselves.
* Report some of the metrics by default.

## Verifying this change

Unit tests have been added to test the metrics reporting and additional information passing between `SplitReader` and `RecordEmitter`.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (yes)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (yes)
  - If yes, how is the feature documented? (docs / JavaDocs)

becketqin commented:
@StephanEwen @zentol This is the patch for FLIP-33. Do you have time to take a look? Thanks!

StephanEwen commented:
Sorry to be late to the game here, but could you share a bit more information on what the original setup was?
Specifically, what was your checkpoint storage system that offered such bad stream read performance? Was it HDFS? OSS? S3?

Looking at this change here, it seems very big (40 files) for "just" introducing a buffer in a stream.
So I tend to be -1 on the change as it is.

Two other options to solve this:

(1) Input stream buffering is a property of the `CheckpointStorage`. It is created there, rather than in the state backends that have to wrap the stream.

(2) Alternatively, we can make it a contract that all FileSystem implementations return well buffered streams. Some already do this by default, wrapping them with another buffered stream adds just another layer and extra copying of bytes, costing performance.

At a first glance, I'd say let's go with option (2) if possible, otherwise option (1).
Hence the question: Which FS did you use that had such bad performing streams?

StephanEwen commented:
Taking a look now.

@becketqin The code currently fails with `NullPointerException` in some tests.

StephanEwen commented:
Thanks for implementing this.

At a first glance, I think moving the `MetricNames` to `flink-core` does not work too well.  The metric names don't have a meaning by themselves. They depend on the Metric Scope (think of it like a package/namespace) to build the full name. Without that, they are like a classname without package. Not unique, not instantiable, not queryable, etc. Just an incomplete fragment. They cannot be used in islolation.
That was more coherent in the original place, the runtime metrics package, next to the runtime metric groups, which define the scope and build the full name.

This looks like "premature consolidation". We could just keep the runtime metric names as they are, and have a separate class for source metric names. A bit like the config options, where the connector base also defines it's options in it's own module, rather than in a shared core class. The argument for metrics names is even stronger, because of the aforementioned issue that the names alone are incomplete.

StephanEwen commented:
@zentol Curious what your take is here.

StephanEwen commented:
The PR adds a new `numBytesIn` metric to the `OperatorIOMetricGroup`. That means all operators now show that metric. But the metric is zero for the majority of operators (all except for sources), because all other operators never receive bytes (just records).
I think that is an issue.

An idea would be investigate the following approach:
  - Define a `SourceMetricGroup` interface that extends `MetricGroup` and has the additional methods for pre-defined metrics (`getNumRecordsIn()`, ...). That would also make the `SourceContext` interface a bit nicer, because all metric-related methods are in the metrics group. We don't need to extend the `SourceContext` interface each time a new pre-defined metric gets added.

  - Implement a `SourceMetricGroupImpl` that reuses some metrics (counters / Gauges) from the `OperatorIOMetricGroup` (the `numRecordsIn`) and some metrics from the `TaskIOMetricGroup` (the `numBytesIn` metrics). That way the source's `numBytesIn` contribute to the tasks's `numBytesIn` and are also displayed in the web UI main job overview.

StephanEwen commented:
A thought on performance: There are some metrics that use `volatile` variables.

So far we always avoided that. Even "infrequently accessed" fields, like watermarks, are quite frequent in some use cases.
And a single volatile write has a very significant overhead in a high-performance pipeline. It is the number one thing we try to avoid, because it breaks the processor pipeline and puts pressure on the memory bus and processor's cache coherency mechanism. For the metrics, performance (least impact on main data paths) should have clear precedence.

For metrics, we don't need a hard "happens before" relationship or any "immediate visibility". If the visibility is some 100s of milliseconds later (which would be a lot), it is no issue for the metrics at all, which are periodically reported every few seconds.

The case where the JIT inlines the variable and eliminates cross-thread visibility entirely is AFAIK not realistic here. And if we ever find that the metric gets not reported, we can still use an occasional volatile write (like we do occasional volatile exception checks, every 1000 records, in the network stack).

becketqin commented:
> A thought on performance: There are some metrics that use volatile variables.
> 
> So far we always avoided that. Even "infrequently accessed" fields, like watermarks, are quite frequent in some use cases.
> And a single volatile write has a very significant overhead in a high-performance pipeline. It is the number one thing we try to avoid, because it breaks the processor pipeline and puts pressure on the memory bus and processor's cache coherency mechanism. For the metrics, performance (least impact on main data paths) should have clear precedence.
> 
> For metrics, we don't need a hard "happens before" relationship or any "immediate visibility". If the visibility is some 100s of milliseconds later (which would be a lot), it is no issue for the metrics at all, which are periodically reported every few seconds.
> 
> The case where the JIT inlines the variable and eliminates cross-thread visibility entirely is AFAIK not realistic here. And if we ever find that the metric gets not reported, we can still use an occasional volatile write (like we do occasional volatile exception checks, every 1000 records, in the network stack).

Thanks for the comments and suggestions. I was actually struggling on this a little bit. The current implementation is based on the following assumptions:
* The volatile read is usually as cheap as a normal variable read unless there is write contention. Modern CPUs can achieve a cheap read if there isn't a contention on the shared variable. The CPU cache line usually can cache the shared variable and just read from it. The reads only go to main memory if the value has been updated by another core, in that case the reading CPU's cache line is invalidated.
* In the current implementation, the main thread performs a read on a volatile variable of `shouldUpdateCurrentEventTimeLag` for every record. The `shouldUpdateCurrentEventTimeLag` is set to true by a separate thread every second. If the main thread sees the value becomes true, it will update the volatile variable of `currentEmitEventTimeLag` with `System.currentTimeMillis() - timestamp`. So this is a volatile variable write and potentially a system call that the main thread performs once per second.
* The `watermark` update is infrequent. From your comment this assumption seems to optimistic. So we can probably also update the `currentWatermark` once per second.

Another part that took me some time to think about is whether we should use number-of-record-based interval v.s. time-based interval for metric reporting. In cases when there is a large throughput, these two approaches probably won't make any difference. But in case of low throughput, time-based-interval seems preferable. Imagine we emit record every 1000 records, but there is a low throughput stream that only processes one record every 5 seconds. In order to see the metrics get updated we need to wait for over an hour, which seems bad.

I am curious do we have some experience that invalidates these assumptions?

becketqin commented:
@StephanEwen Thanks for the comment. I have updated the patch. Would you help take a look? 

I have also left some thoughts regarding the performance impact. Please let me know if you still have concerns. 

Just want to add to the performance discussion. The first version of implementation I had has a `DynamicMetricSampler` class which does the following:
1. Take a target metric reporting overhead in percentage, for example, 0.01%.  
2. Measure the absolute time it takes to report a metric, e.g. 1000 ns. 
3. Based on the overhead and throughput, it calculates the metric sampling interval. In the above case, if the throughput is 1000 records per second, each record takes 1 ms (1,000,000 ns) to process. If the overhead is 0.01%, the budget for metric reporting is 100 ns per record. Given that each metric reporting takes 1000 ns, the sampling interval should be every 10 records.
4. The metric sampling interval is adjusted periodically to reflect the latest throughput.

The above logic allows a quantifiable bounded performance impact on the throughput. But I removed because In most cases, periodical reporting is good enough, e.g. reporting metrics every second. So we can avoid some complexity. If you think this approach helps, we can also bring that in.

StephanEwen commented:
## On Volatile Reads and Performance Considerations

I think `volatile` read assumption is completely correct. There is a common interpretation that this just means "load variable from memory if modified by another thread", meaning it's for free if no other thread touches it. But volatile reads in Java insert a "load barrier" for the processor - it affect all variables and memory after that barrier, to guarantee the happens-before relationship between everything before/after that volatile read. In hardware (x86), a load barrier means draining the processor's load buffer completely before continuing, which may (and frequently will) stall the processor pipeline.

As a concrete example in Flink: If you look at the current mailbox runtime in the StreamTask, it does in fact a lot more work before processing a record than the previous "run loop", but is still faster, purely because it has less volatile variable access.

Here are two good references:
  - https://mechanical-sympathy.blogspot.com/2011/07/memory-barriersfences.html
  - https://www.infoq.com/articles/memory_barriers_jvm_concurrency/

The conclusion here is to use volatile only where you want to establish load/store barriers. Typically at the end of a block of work.
In Flink this means the following:
  - We have the load/store barriers already once after every record is processed, by the mailbox. That is enough to make sure changes by the processing thread during one record's processing are made visible to other threads before the next record is processed.
  - We don't need a specific additional happens-before relationship around values for metrics, so the metrics do not need an additional volatile access.

There is now also an additional static thread to work with that variable, which is one more utility thread. There have been quite a few of those recently added to Flink, and I think we should start to be careful about adding those. One thread alone doesn't hurt, but it adds up over time. Every thread adds memory footprint overhead (more than 1MB, again sounds not like much, but doing this 10s of times makes a difference) and increases overhead through context switches (having 10s of additional threads that jump in every second will eventually be measurable). So threads should be added is necessary, but not if there is another equally good solution.

I think we really should go the "keep it simple" approach here. No volatile on the metric path, no extra threads for "black magic". Simplest possible solution, as predictability as possible. I believe it will actually work well, the remaining metric code follows that principle and I am not aware of any problem.
If we actually discover problems, we can still look at adding a complex magic solution. But I am very hesitant to add that complexity now, without really knowing there is a problems that warrants such a solution. Complexity should not come before the problem.

## Event-time Lag Gauge

We can grab the watermark "for free" this way:
  - We don't have to eagerly compute the lag. The best point to compute it is directly when the metric is probed, in the `Gauge.getValue()` method. The we compute it exactly as often as it is needed.
  - The computation then happens in the metric reporter or the single metric thread that periodically gets all metric values to report them.
  - All watermarks leave the source through the `WatermarkOutput` defined here: https://github.com/apache/flink/blob/master/flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/source/ProgressiveTimestampsAndWatermarks.java#L101
  - The field `WatermarkToDataOutput.maxWatermarkSoFar` already has the value we need. So the `getValue()` method could simply to `Math.max(0L, System.currentTimeMillis() - output.getMaxWatermarkSoFar())`.

Not using a volatile field here would still guarantee visibility (through the barriers in the mailbox), visibility might just be a few micoseconds late, because we don't flush and wait for the store buffer in the processor.

becketqin commented:
@StephanEwen Thanks for the detail explanation. I updated the patch to address your comments and removed the controversial part. Now the patch only contains two commits:

1. Add metric definitions of Source.
    - Expose the Source metrics via `SourceMetricGroup` and `SourceReaderContext`.
    - Let `SourceOperator` report the WatermarkLag.
2. Report numRecordsIn, numRecordsInPerSecond and SourceIdleTime in SourceReaderBase.

Would you help take another look? Thanks.

becketqin commented:
I don't particularly like the current way of passing in the `numBytesInCounter` from the `TaskMetricGroup`. It is probably cleaner to replace that with the `SumCounter`.

becketqin commented:
It is actually even simpler. The `numBytesInCounter` in the `TaskMetricGroup` can be accessed all the way up from the SourceOperator.

StephanEwen commented on flink-runtime/src/main/java/org/apache/flink/runtime/metrics/groups/SourceMetricGroupImpl.java:
```diff
@@ -0,0 +1,74 @@
+	private final Supplier<Long> currentWatermarkSupplier;
+
+	public SourceMetricGroupImpl(
+			OperatorMetricGroup parentMetricGroup,
```
Maybe we can rename this to `operatorMetricGroup` or so, because it is the same metric group, just extended via the proxy. Whenever I read "parent" in the code below, I thought this was the actual parent (the TaskMetricGroup) and though we register the metrics to the wrong group.

StephanEwen commented on flink-streaming-java/src/test/java/org/apache/flink/streaming/api/operators/source/TestingSourceOperator.java:
```diff
@@ -86,6 +87,7 @@ public TestingSourceOperator(
 		this.parallelism = parallelism;
 		this.metrics = UnregisteredMetricGroups.createUnregisteredOperatorMetricGroup();
 		this.readerCreated = false;
+		this.sourceMetricGroup = new TestingSourceMetricGroup();
```
What do you think about having a testing constructor in the `SourceOperator` that accepts a metric group, rather making a parent field public and mutating it from here?

StephanEwen reviewed (commented):
The first commit looks good, +1 to merge the first commit.

StephanEwen commented on flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/source/reader/SourceReaderBase.java:
```diff
@@ -114,16 +137,20 @@ public InputStatus pollNext(ReaderOutput<T> output) throws Exception {
 		if (recordsWithSplitId == null) {
 			recordsWithSplitId = getNextFetch(output);
 			if (recordsWithSplitId == null) {
+				idleSince = clock.absoluteTimeMillis();
```
For measuring intervals, relative time (which is stable) seems like a good choice.

StephanEwen reviewed (commented):
The second commit looks also fine, modulo minor comments.

The main thing I am not sure about is the "idle-time metric".

  - What does the idle-time metric (implemented as it is here) give us, compared to the operator idle time? The difference is rather small, there are few cases where the mailbox is busy even though the source is idle (like firing processing time timers).

  - If we keep this metric, can we find a way to deactivate this during batch processing?

becketqin commented:
@StephanEwen Thanks a lot for the review. I have addressed most of them.

Regarding the questions about sourceIdleTime.
>Can we deactivate this during batch processing? It seems less valuable there, and we have been very performance conscious for batch workloads.
It seems that the cost of skipping updating this metric could be the same or even more than just updating the idle time which is a local variable, because the skipping code would likely result in an if-else branch that involves reading from a variable and branch prediction logic in the CPU.

One change I made in the last update is putting the `sourceIdleTime = -1L` right after we got a non-null recordsWithSplitIds. So the sourceIdleTime will only be updated once per batch, instead of once per record. I think this should have little performance impact on the current code, if any.

>The metric measures the time since last record. That metric fluctuates heavily, resets to zero multiple times per second. A lot of the actual information (how much idleness there was in total) will be lost by the time the metric is sampled. The mailbox idleness tries to measure the average idleness fraction of the total processing time. Is that something that would be good for this metric as well, or should it stay as it is?
I think the idleness fraction is particularly useful to measure the workload of a thread. In the source case, users usually care more about for how long there is no data processed from a source. I don't think this metric is critical, but could help clarify the latency issues in some cases, especially when some connectors do not emit `fetchEventTimeLag` by default. In that case, when the `emitEventTimeLag` reports a large value, we need to know whether it is because Flink was not able to get the data from the external system, or it is because the `RecordEmitter` took long to process the record. The `sourceIdleTime` would be helpful in that case.

becketqin commented:
@rmetzger @dianfu I would like to get your opinions about getting this PR into Flink 1.12 release. 

This PR adds FLIP-33 metrics support to the Source. The reasons I think we should have it in 1.12 are following:
1. It is the last major piece that would make the Source usable in serious case - people won't be able to actually use a connector in production if there is no metric. 
2. If the individual connectors implement the metrics by themselves. They may find those metrics clash with the metrics that we add later.
3. FLIP-27 is still a relatively new feature and the patch only touches the Source. So the risk of introducing problem is low.

What do you think? Thanks.

dianfu commented:
@becketqin Fair enough for me. +1 from my side.

rmetzger commented:
I generally think it is problematic to merge new features almost a week after the feature freeze. If everybody would do this, then the release would drag on forever. In particular if a change is causing build instabilities which block the release process.

However in this case, I see the feature is mostly limited to connectors (thus we can revert it in case of instabilities), and it seems that testing of the Kafka connector hasn't started yet either. I'd be okay with merging it, given the user benefits.

rmetzger commented:
I just looked over the review comments so far again, and I also realized that the code doesn't even compile at this point. I'm not sure if it is really a good idea to merge this still for the 1.12 release, as it could lead to instabilities that will be difficult to debug, or API choices that will be hard to change in future releases.

becketqin commented:
@rmetzger Thanks for sharing your thoughts and concerns. I understand your concerns about merging new features after feature freeze. And I am fine with leaving this patch out of 1.12.

In case it might relieve your concern a little bit, the reason this patch did not compile was due to a race condition with another PR. And I kind of think that compared with other patches in the source, this one is probably least likely to cause instability given that it just reports metrics.

becketqin commented:
@StephanEwen As we discussed offline, I updated the patch to remove reporting of `sourceIdleTime`.  The only metric reported out of the box now is `numRecordsIn`.
